src/cookies.js

```diff
@@ -83,29 +90,74 @@ function tryGetDocumentCookieNoInline(win) {
  *     - highestAvailableDomain: If true, set the cookie at the widest domain
  *       scope allowed by the browser. E.g. on example.com if we are currently
  *       on www.example.com.
- *     - domain: Explicit domain to set.
+ *     - domain: Explicit domain to set. domain overrides HigestAvailableDomain
  *     - allowOnProxyOrigin: Allow setting a cookie on the AMP Cache.
  */
 export function setCookie(win, name, value, expirationTime, opt_options) {
   checkOriginForSettingCookie(win, opt_options, name);
-  if (opt_options && opt_options.highestAvailableDomain) {
-    const parts = win.location.hostname.split('.');
-    let domain = parts[parts.length - 1];
-    for (let i = parts.length - 2; i >= 0; i--) {
-      domain = parts[i] + '.' + domain;
-      trySetCookie(win, name, value, expirationTime, domain);
-      if (getCookie(win, name) == value) {
-        return;
-      }
-    }
-  }
   let domain = undefined;
+  // Respect explicitily set domain over higestAvailabeDomain
   if (opt_options && opt_options.domain) {
     domain = opt_options.domain;
+  } else if (opt_options && opt_options.highestAvailableDomain) {
+    domain = getHighestAvailableDomain(win);
   }
   trySetCookie(win, name, value, expirationTime, domain);
 }
 
+/**
+ * Attemp to find the HighestAvailableDomain on
+ * @param {!Window} win
+ */
+export function getHighestAvailableDomain(win) {
+  // <meta name='amp-cookie-scope'>. Need to respect the meta first.
+  const metaTag = win.document.head.querySelector(
+    "meta[name='amp-cookie-scope']"
+  );
+
+  if (metaTag) {
+    // The content value could be an empty string. Return null instead
+    const cookieScope = metaTag.getAttribute('content') || '';
+    // Verify the validness of the amp-cookie-scope meta value
+    const sourceOrigin = getSourceOrigin(win.location.href);
+    // Verify the meta tag content value is valid
+    if (sourceOrigin.endsWith('.' + cookieScope)) {
+      return cookieScope;
+    } else {
+      // When the amp-cookie-scope value is invalid, fallback to the exact origin
+      // the document is contained in.
+      // sourceOrigin in the format of 'https://xxx or http://xxx'
+      return sourceOrigin.split('://')[1];
+    }
+  }
+
+  if (isProxyOrigin(win.location.href)) {
```

Comment:
I understand this is easier to read. But I'm afraid it also add a few more bytes. I don't know if compiler is smart enough to combine the two `return null` to one.

Reply:
agree. let's do `!isProxyOrigin`

Reply:
done